junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/AnnotatedElementContext.java

```diff
@@ -0,0 +1,109 @@
+ */
+@API(status = EXPERIMENTAL, since = "5.10")
+public interface AnnotatedElementContext {
+
+	/**
+	 * Get the {@link AnnotatedElement} for this context.
+	 *
+	 * <h4>WARNING</h4>
+	 * <p>When searching for annotations on the annotated element in this context,
+	 * favor {@link #isAnnotated(Class)}, {@link #findAnnotation(Class)}, and
+	 * {@link #findRepeatableAnnotations(Class)} over methods in the
+	 * {@link AnnotatedElement} API due to a bug in {@code javac} on JDK versions prior
+	 * to JDK 9.
+	 *
+	 * @return the annotated element; never {@code null}
+	 */
+	AnnotatedElement getAnnotatedElement();
+
+	/**
+	 * Determine if an annotation of {@code annotationType} is either
+	 * <em>present</em> or <em>meta-present</em> on the {@link AnnotatedElement} for
+	 * this context.
+	 *
+	 * <h4>WARNING</h4>
+	 * <p>Favor the use of this method over directly invoking
+	 * {@link AnnotatedElement#isAnnotationPresent(Class)} due to a bug in {@code javac}
+	 * on JDK versions prior to JDK 9.
+	 *
+	 * @param annotationType the annotation type to search for; never {@code null}
+	 * @return {@code true} if the annotation is present or meta-present
+	 * @see #findAnnotation(Class)
+	 * @see #findRepeatableAnnotations(Class)
+	 */
+	boolean isAnnotated(Class<? extends Annotation> annotationType);
+
+	/**
+	 * Find the first annotation of {@code annotationType} that is either
+	 * <em>present</em> or <em>meta-present</em> on the {@link AnnotatedElement} for
+	 * this context.
+	 *
+	 * <h4>WARNING</h4>
+	 * <p>Favor the use of this method over directly invoking annotation lookup
+	 * methods in the {@link AnnotatedElement} API due to a bug in {@code javac} on JDK
+	 * versions prior to JDK 9.
+	 *
+	 * @param <A> the annotation type
+	 * @param annotationType the annotation type to search for; never {@code null}
+	 * @return an {@code Optional} containing the annotation; never {@code null} but
+	 * potentially empty
+	 * @see #isAnnotated(Class)
+	 * @see #findRepeatableAnnotations(Class)
+	 */
+	<A extends Annotation> Optional<A> findAnnotation(Class<A> annotationType);
+
+	/**
+	 * Find all <em>repeatable</em> {@linkplain Annotation annotations} of
+	 * {@code annotationType} that are either <em>present</em> or
+	 * <em>meta-present</em> on the {@link AnnotatedElement} for this context.
+	 *
+	 * <h4>WARNING</h4>
+	 * <p>Favor the use of this method over directly invoking annotation lookup
+	 * methods in the {@link AnnotatedElement} API due to a bug in {@code javac} on JDK
+	 * versions prior to JDK 9.
+	 *
+	 * @param <A> the annotation type
+	 * @param annotationType the repeatable annotation type to search for; never
+	 * {@code null}
+	 * @return the list of all such annotations found; neither {@code null} nor
+	 * mutable, but potentially empty
+	 * @see #isAnnotated(Class)
+	 * @see #findAnnotation(Class)
+	 * @see java.lang.annotation.Repeatable
+	 */
+	<A extends Annotation> List<A> findRepeatableAnnotations(Class<A> annotationType);
```

Comment:
Actually, now that I think about it, all of the last 3 methods could be `default` methods that delegate to the corresponding `AnnotatedElement`-accepting methods in `AnnotationUtils`, passing the value of `getAnnotatedElement()`.

Only `DefaultParameterContext` needs to override those to use the special `Parameter`-accepting variants of those utility methods, which it already does.

It's not a big deal, but it might make sense... if we later end up using `AnnotatedElementContext` for other use cases.

Reply:
I assume I should also remove these methods from `ParameterContext` so https://github.com/junit-team/junit5/pull/3394#discussion_r1264394562 would no longer be applicable.

For these methods in `AnnotatedElementContext`, what values should I use for `@since` and `@API`? Same as @marcphilipp's proposal?

Reply:
I'd like to keep the declarations in `ParameterContext` for documentation purposes. We can make them `default`, though, and delegate to their `super` implementations.